tasks/broad/Alignment.wdl

```diff
@@ -17,6 +17,86 @@ version 1.0
 
 import "../../structs/dna_seq/DNASeqStructs.wdl"
 
+task Bazam {
+  input {
+    File input_bam
+    File input_bam_index
+    String sample_name
+    String output_bam_basename
+    String duplicate_metrics_fname
+
+    # reference_fasta.ref_alt is the .alt file from bwa-kit
+    # (https://github.com/lh3/bwa/tree/master/bwakit),
+    # listing the reference contigs that are "alternative".
+    ReferenceFasta reference_fasta
+
+    Int preemptible_tries
+    Boolean to_cram = false
+    String? subset_region 
+  }
+  
+  String output_format = if to_cram then "cram" else "bam"
+  String bazam_regions = if defined(subset_region) then "--regions ~{subset_region} " else ""
+  
+  Int bwa_cpu = 20
+  Int bazam_cpu = 6
+  Int bamsormadup_cpu = 6
+  Int total_cpu = bwa_cpu + bazam_cpu + bamsormadup_cpu
+
+  String rg_line = "@RG\\tID:~{sample_name}\\tSM:~{sample_name}"
+  
+  String output_file = if to_cram then "~{output_bam_basename}.cram" else "~{output_bam_basename}.bam"
+  String output_index = if to_cram then "~{output_bam_basename}.crai" else "~{output_bam_basename}.bai"
+  
+  # BWA command options:
+  # -K      process INT input bases in each batch regardless of nThreads (for reproducibility)
+  # -p      smart pairing (ignoring in2.fq)
+  # -v 3    minimum score to output [30]
+  # -t 16   threads
+  # -Y      use soft clipping for supplementary alignments
+  # -R      read group header line such as '@RG\tID:foo\tSM:bar'
+  # -M      mark shorter split hits as secondary
+  command <<<
+    set -o pipefail
+    set -ex
+
+    (while true; do df -h; pwd; du -sh *; sleep 300; done) &
+    
+    bazam -Xmx16g -Dsamjdk.reference_fasta=~{reference_fasta.ref_fasta} \
+      ~{bazam_regions} -n~{bazam_cpu} -bam ~{input_bam} | \
+    bwa mem -K 100000000 -p -v3 -t~{bwa_cpu} -Y -R '~{rg_line}' \
+      ~{reference_fasta.ref_fasta} /dev/stdin - \
+      2> >(tee ~{output_bam_basename}.bwa.stderr.log >&2) | \
+    bamsormadup inputformat=sam threads=~{bamsormadup_cpu} SO=coordinate \
+      M=~{duplicate_metrics_fname} \
+      outputformat=sam | \
+    samtools view -T ~{reference_fasta.ref_fasta} \
+      -O ~{output_format} \
+      -o ~{output_file}
+    
+    samtools index -@~{total_cpu} ~{output_file} ~{output_index}
+
+    df -h; pwd; du -sh *
+  >>>
+  
+  runtime {
+    # docker: "australia-southeast1-docker.pkg.dev/fewgenomes/images/bazam:v2"
+    # cromwell doesn't work with artifact registry:
+    # java.lang.Exception: Registry australia-southeast1-docker.pkg.dev is not supported
+    docker: "gcr.io/fewgenomes/bazam:v2"
+    preemptible: preemptible_tries
+    memory: "64 GiB"
+    cpu: total_cpu
+    disks: "local-disk " + 500 + " HDD"
```

Comment:
Nice -- this seems more robust than a multiplier based on BAM vs CRAM and is probably only slightly more expensive.

Reply:
Actually reduced to 300G, seems to be enough.